File: Demo/DJISdkDemo/DJISdkDemo/Demo/Navigation/IOC/NavigationIOCViewController.m
//
//  DJIIOCViewController.m
//  DJISdkDemo
//
//  Created by Ares on 15/7/1.
//  Copyright (c) 2015年 DJI. All rights reserved.
//

#import "NavigationIOCViewController.h"
#import "VideoPreviewer.h"
#import "DJIDemoHelper.h"
#import "DJIAircraftAnnotation.h"
#import "DJIAircraftAnnotationView.h"
#import "DJIWaypointAnnotation.h"
#import "DJIWaypointAnnotationView.h"

@interface NavigationIOCViewController ()
@property (weak, nonatomic) IBOutlet UISegmentedControl *iocTypeSegmentCtrl;
@property (weak, nonatomic) IBOutlet MKMapView *mapView;
@property (weak, nonatomic) IBOutlet UIView *previewView;
- (IBAction)onEnterNavigationButtonClicked:(id)sender;
- (IBAction)onStartIOCButtonClicked:(id)sender;
- (IBAction)onStopIOCButtonClicked:(id)sender;
- (IBAction)onRecordButtonClicked:(id)sender;
- (IBAction)onLockCourseButtonClicked:(id)sender;

@property(nonatomic, strong) DJIAircraftAnnotation* aircraftAnnotation;
@property(nonatomic, strong) DJIWaypointAnnotation* waypointAnnotation;

@property(nonatomic, assign) DJIDroneType droneType;

@end

@implementation NavigationIOCViewController

-(id) initWithDroneType:(DJIDroneType)type
{
    self = [super init];
    if (self) {
        self.droneType = type;
        

    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    
    if (self.connectedDrone) {
        self.drone = self.connectedDrone;
    }
    else
    {
        self.drone = [[DJIDrone alloc] initWithType:self.droneType];
    }
    
    self.drone.delegate = self;
    self.drone.mainController.mcDelegate = self;
    self.drone.camera.delegate = self;
    self.navigationManager = self.drone.mainController.navigationManager;
    self.navigationManager.delegate = self;
    
    for (int i = 100; i < 105; i++) {
        UIButton* btn = (UIButton*)[self.view viewWithTag:i];
        if (btn) {
            btn.layer.cornerRadius = btn.frame.size.width * 0.5;
            btn.layer.borderWidth = 1.2;
            btn.layer.borderColor = [UIColor redColor].CGColor;
            btn.layer.masksToBounds = YES;
        }
    }
    

    
    [[VideoPreviewer instance] start];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

-(void) viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [[VideoPreviewer instance] setView:self.previewView];
    [[VideoPreviewer instance] setDecoderDataSource:[self dataSourceFromDroneType:_drone.droneType]];
    self.mapView.delegate = self;
    
    [self.drone connectToDrone];
    [self.drone.mainController startUpdateMCSystemState];
    [self.drone.camera startCameraSystemStateUpdates];
}

-(void) viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    self.mapView.delegate = nil;
    
    [self.drone.mainController stopUpdateMCSystemState];
    [self.drone.camera stopCameraSystemStateUpdates];
    [self.drone disconnectToDrone];
    self.drone.delegate = nil;
    
    [[VideoPreviewer instance] unSetView];
}

- (IBAction)onEnterNavigationButtonClicked:(id)sender
{
    [self.navigationManager enterNavigationModeWithResult:^(DJIError *error) {
        ShowResult(@"Enter Navigation:%@", error.errorDescription);
    }];
}

- (IBAction)onStartIOCButtonClicked:(id)sender
{
    DJIIOCType type = (DJIIOCType)(self.iocTypeSegmentCtrl.selectedSegmentIndex + 1);
    [self.navigationManager.iocMission setIocType:type];
    [self.navigationManager.iocMission startMissionWithResult:^(DJIError *error) {
        ShowResult(@"Start IOC:%@", error.errorDescription);
    }];
}

- (IBAction)onStopIOCButtonClicked:(id)sender
{
    [self.navigationManager.iocMission stopMissionWithResult:^(DJIError *error) {
        ShowResult(@"Stop IOC:%@", error.errorDescription);
    }];
}

- (IBAction)onRecordButtonClicked:(id)sender {
    if (self.isRecording) {
        [self.drone.camera stopRecord:^(DJIError *error) {
            ShowResult(@"Stop Record:%@", error.errorDescription);
        }];
    }
    else
    {
        [self.drone.camera startRecord:^(DJIError *error) {
            ShowResult(@"Start Record:%@", error.errorDescription);
        }];
    }
    
}

- (IBAction)onLockCourseButtonClicked:(id)sender
{
    [self.navigationManager.iocMission lockCourseUsingCurrentDirectionWithResult:^(DJIError *error) {
        ShowResult(@"Lock Course:%@", error.errorDescription);
    }];
}

#pragma mark -

-(void) droneOnConnectionStatusChanged:(DJIConnectionStatus)status
{
    
}

#pragma mark - 

-(void) camera:(DJICamera*)camera didReceivedVideoData:(uint8_t*)videoBuffer length:(int)length
{
    uint8_t* pBuffer = (uint8_t*)malloc(length);
    memcpy(pBuffer, videoBuffer, length);
    [[[VideoPreviewer instance] dataQueue] push:pBuffer length:length];
}

-(void) camera:(DJICamera*)camera didUpdateSystemState:(DJICameraSystemState*)systemState
{
    if (systemState.workMode != CameraWorkModeRecord) {
        if (self.drone.droneType == DJIDrone_Inspire) {
            [(DJIInspireCamera*)self.drone.camera setCameraWorkMode:CameraWorkModeRecord withResult:nil];
        }
        else if (self.drone.droneType == DJIDrone_Phantom3Advanced)
        {
            [(DJIPhantom3AdvancedCamera*)self.drone.camera setCameraWorkMode:CameraWorkModeRecord withResult:nil];
        }
        else if (self.drone.droneType == DJIDrone_Phantom3Professional)
        {
            [(DJIPhantom3ProCamera*)self.drone.camera setCameraWorkMode:CameraWorkModeRecord withResult:nil];
        }
        
    }
    if (self.isRecording != systemState.isRecording) {
        UIButton* recBtn = (UIButton*)[self.view viewWithTag:103];
        if (recBtn) {
            [recBtn setTitleColor:(systemState.isRecording ? [UIColor redColor] : [UIColor blackColor]) forState:UIControlStateNormal];
        }
        self.isRecording = systemState.isRecording;
    }
}

#pragma mark -

-(void) mainController:(DJIMainController*)mc didUpdateSystemState:(DJIMCSystemState*)state
{
    if (CLLocationCoordinate2DIsValid(state.droneLocation)) {
        if (self.aircraftAnnotation == nil) {
            self.aircraftAnnotation = [[DJIAircraftAnnotation alloc] initWithCoordiante:state.droneLocation];
            [self.mapView addAnnotation:self.aircraftAnnotation];
            
            MKCoordinateRegion region = {0};
            region.center = state.droneLocation;
            region.span.latitudeDelta = 0.001;
            region.span.longitudeDelta = 0.001;
            
            [self.mapView setRegion:region animated:YES];
        }
        [self.aircraftAnnotation setCoordinate:state.droneLocation];
        double heading = RADIAN(state.attitude.yaw);
        DJIAircraftAnnotationView* annoView = (DJIAircraftAnnotationView*)[self.mapView viewForAnnotation:self.aircraftAnnotation];
        [annoView updateHeading:heading];
    }
    if (CLLocationCoordinate2DIsValid(state.homeLocation)) {
        if (self.waypointAnnotation == nil) {
            self.waypointAnnotation = [[DJIWaypointAnnotation alloc] initWithCoordiante:state.homeLocation];
            [self.mapView addAnnotation:self.waypointAnnotation];
        }
        
        [self.waypointAnnotation setCoordinate:state.homeLocation];
    }
}

#pragma mark - DJINavigationDelegate

-(void) onNavigationMissionStatusChanged:(DJINavigationMissionStatus*)missionStatus
{
    if (missionStatus.missionType == DJINavigationMissionIOC) {
        DJIIOCMissionStatus* iocStatus = (DJIIOCMissionStatus*)missionStatus;
        if (iocStatus.iocType == DJIIOCTypeCourseLock) {

        }
        else
        {
            
        }
    }
}

#pragma mark - MKMapViewDelegate

- (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotation:(id <MKAnnotation>)annotation
{
    if ([annotation isKindOfClass:[DJIAircraftAnnotation class]]) {
        MKAnnotationView* annoView = [self.mapView dequeueReusableAnnotationViewWithIdentifier:@"DJIAircraftAnnotationView"];
        if (annoView == nil) {
            annoView = [[DJIAircraftAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:@"DJIAircraftAnnotationView"];
        }
        return annoView;
    }
    else if ([annotation isKindOfClass:[DJIWaypointAnnotation class]]) {
        DJIWaypointAnnotationView* annoView = (DJIWaypointAnnotationView*)[self.mapView dequeueReusableAnnotationViewWithIdentifier:@"DJIWaypointAnnotationView"];
        if (annoView == nil) {
            annoView = [[DJIWaypointAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:@"DJIWaypointAnnotationView"];
            annoView.titleLabel.text = @"H";
            return annoView;
        }
    }
    
    return nil;
}
@end

File: Sample Code/SwiftSampleCode/DJISDKSwiftDemo/Missions/IOC/NavigationIOCViewController.swift
//
//  NavigationIOCViewController.swift
//  DJISDKSwiftDemo
//
//  Created by DJI on 15/7/1.
//  Copyright (c) 2015 DJI. All rights reserved.
//
import UIKit
import DJISDK
import MapKit
class NavigationIOCViewController: DJIBaseViewController, DJIFlightControllerDelegate, DJICameraDelegate, MKMapViewDelegate {

    @IBOutlet weak var iocTypeSegmentCtrl: UISegmentedControl!
    @IBOutlet weak var mapView: MKMapView!
    @IBOutlet weak var previewView: UIView!
    var djiMapView: DJIMapView? = nil
    var isRecording: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        let aircraft: DJIAircraft? = self.fetchAircraft()
        if aircraft != nil {
            // Do any additional setup after loading the view from its nib.
            aircraft!.flightController?.delegate = self

        }
        for i in 100 ..< 105 {
            let btn: UIButton? = self.view!.viewWithTag(i) as? UIButton
            if btn != nil {
                btn!.layer.cornerRadius = btn!.frame.size.width * 0.5
                btn!.layer.borderWidth = 1.2
                btn!.layer.borderColor = UIColor.red.cgColor
                btn!.layer.masksToBounds = true
            }
        }
        self.djiMapView = DJIMapView(mapView: self.mapView)
     
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        let aircraft: DJIAircraft? = self.fetchAircraft()
        if aircraft != nil {
            aircraft!.flightController?.delegate = self
        }
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        let aircraft: DJIAircraft? = self.fetchAircraft()
        if aircraft != nil {
            if aircraft!.flightController?.delegate === self {
                aircraft!.flightController!.delegate = nil
            }
        }
        
        if djiMapView != nil  {
            self.djiMapView = nil
        }
    }

    @IBAction func onStartIOCButtonClicked(_ sender: AnyObject) {
        let aircraft: DJIAircraft? = self.fetchAircraft()
        if aircraft != nil {
            
        let type: DJIFlightOrientationMode = DJIFlightOrientationMode(rawValue: UInt8(self.iocTypeSegmentCtrl.selectedSegmentIndex))!
        aircraft!.flightController?.setFlightOrientationMode(type, withCompletion: {[weak self] (error: Error?) -> Void in
            if (error != nil){
                self?.showAlertResult("Start IOC: \(error!)")
            }
        })
        }
    }

    @IBAction func onStopIOCButtonClicked(_ sender: AnyObject) {
        let aircraft: DJIAircraft? = self.fetchAircraft()
        if aircraft != nil {
        
        
        aircraft!.flightController?.setFlightOrientationMode(DJIFlightOrientationMode.defaultAircraftHeading, withCompletion: {[weak self] (error: Error?) -> Void in
            if (error != nil) {
                self?.showAlertResult("Stop IOC: \(error!)")
            }
        })
        }
    }

    @IBAction func onRecordButtonClicked(_ sender: AnyObject) {
        let aircraft: DJIAircraft? = self.fetchAircraft()
        if aircraft == nil {
            return
        }
        if self.isRecording {
            aircraft!.camera!.stopRecordVideo(completion: {[weak self] (error: Error?) -> Void in
                if (error != nil ) {
                    self?.showAlertResult("Stop Record: \(error!)")
                } else {
                    self?.showAlertResult("Stop Record:Success")
                }
            })
        }
        else {
            aircraft!.camera!.startRecordVideo(completion: {[weak self] (error: Error?) -> Void in
                if (error != nil ) {
                    self?.showAlertResult("Start Record: \(error!)")
                } else {
                    self?.showAlertResult("Start Record:Success")
                }            })
        }
    }

    @IBAction func onLockCourseButtonClicked(_ sender: AnyObject) {
        let aircraft: DJIAircraft? = self.fetchAircraft()
        if aircraft == nil {
            return
        }
        aircraft!.flightController?.lockCourseUsingCurrentDirection(completion: {[weak self] (error: Error?) -> Void in
            
            if (error != nil ) {
                self?.showAlertResult("Lock Course: \(error!)")
            } else {
                self?.showAlertResult("Lock Course: Success")
            }
        })
    }

  
    func camera(_ camera: DJICamera, didUpdate systemState: DJICameraSystemState) {
        let aircraft: DJIAircraft? = self.fetchAircraft()
        if aircraft == nil {
            return
        }
        if systemState.mode != DJICameraMode.recordVideo {
            aircraft!.camera!.setCameraMode(DJICameraMode.recordVideo, withCompletion: nil)
        }
        if self.isRecording != systemState.isRecording {
            let recBtn: UIButton? = self.view!.viewWithTag(103) as? UIButton
            if recBtn != nil {
                recBtn!.setTitleColor((systemState.isRecording ? UIColor.red : UIColor.black), for: UIControlState())
            }
            self.isRecording = systemState.isRecording
        }
    }

    func flightController(_ fc: DJIFlightController, didUpdateSystemState state: DJIFlightControllerCurrentState) {
        if CLLocationCoordinate2DIsValid(state.aircraftLocation) {
            let heading: Double = state.attitude.yaw*M_PI/180.0
            djiMapView!.updateAircraftLocation(state.aircraftLocation, withHeading: heading)
         
        }
        if CLLocationCoordinate2DIsValid(state.homeLocation) {
            djiMapView!.updateHomeLocation(state.homeLocation)
        }
    }
 
}
